lib/Sema/CSWalker.cpp

```diff
@@ -205,6 +205,159 @@ class LocalCSWalker : public glu::ast::ASTWalker<LocalCSWalker, void> {
+                constraints.push_back(
+                    glu::sema::Constraint::createBindOverload(
+                        _cs.getAllocator(), node->getType(), funcDecl, node
+                    )
+                );
+            }
+        }
+
+        if (!constraints.empty()) {
+            _cs.addConstraint(
+                glu::sema::Constraint::createDisjunction(
+                    _cs.getAllocator(), constraints, node, true
+                )
+            );
+        }
+    }
+
+    /// @brief Processes variable declarations that might contain function
+    /// references
+    void processVariableDeclarations(
+        glu::ast::CallExpr *node, llvm::ArrayRef<glu::ast::DeclBase *> decls
+    )
+    {
+        for (auto decl : decls) {
+            if (auto varLetDecl = llvm::dyn_cast<glu::ast::VarLetDecl>(decl)) {
+                auto declType = varLetDecl->getType();
+                if (!declType)
+                    continue;
+
+                glu::types::FunctionTy *functionTy
+                    = extractFunctionType(declType);
+
+                if (functionTy) {
+                    _cs.addConstraint(
+                        Constraint::createEqual(
+                            _cs.getAllocator(), node->getType(),
+                            functionTy->getReturnType(), node
+                        )
+                    );
+                    constrainArguments(node, functionTy);
+                }
+            }
+        }
+    }
+
+    /// @brief Handles function calls through function pointers
+    void handlePointerCall(glu::ast::CallExpr *node)
+    {
+        auto calleeType = node->getCallee()->getType();
+        if (!calleeType)
+            return;
+
+        glu::types::FunctionTy *functionTy = extractFunctionType(calleeType);
+
+        if (functionTy) {
+            _cs.addConstraint(
+                Constraint::createEqual(
+                    _cs.getAllocator(), node->getType(),
+                    functionTy->getReturnType(), node
+                )
+            );
+            constrainArguments(node, functionTy);
+        } else {
+            // Fallback: create conversion constraint for unknown callee type
+            _cs.addConstraint(
+                Constraint::createConversion(
+                    _cs.getAllocator(), calleeType, node->getType(), node
+                )
+            );
+        }
+    }
+
+    /// @brief Extracts function type from pointer or direct function type
+    glu::types::FunctionTy *extractFunctionType(glu::types::TypeBase *type)
+    {
+        if (auto funcTy = llvm::dyn_cast<glu::types::FunctionTy>(type)) {
+            return funcTy;
+        }
+
+        if (auto ptrTy = llvm::dyn_cast<glu::types::PointerTy>(type)) {
+            return llvm::dyn_cast<glu::types::FunctionTy>(ptrTy->getPointee());
+        }
+
+        return nullptr;
+    }
+
+    /// @brief Creates constraints for function call arguments
+    void constrainArguments(
```

Comment:
this shouldn’t be done here, the only constraint you should have is function to function. Argument etc is handled in the constraint simplification step of the solver.

Reply:
this should be deleted